src/toplevel.c

```diff
@@ -267,25 +202,44 @@ jl_value_t *jl_eval_module_expr(jl_module_t *parent_module, jl_expr_t *ex)
     }
 #endif
 
-    arraylist_push(&module_stack, newm);
-
-    if (outermost == NULL || parent_module == jl_main_module) {
-        JL_TRY {
-            size_t i, l = module_stack.len;
-            for (i = stackidx; i < l; i++) {
-                jl_module_t *m = (jl_module_t*)module_stack.items[i];
+    uintptr_t *refcnt = (uintptr_t*)ptrhash_bp(&jl_current_modules, (void*)newm);
+    assert(*refcnt > (uintptr_t)HT_NOTFOUND);
+    *refcnt -= 1;
+    // newm should be reachable from somewhere else by now
+
+    if (jl_module_init_order == NULL)
+        jl_module_init_order = jl_alloc_vec_any(0);
+    jl_array_ptr_1d_push(jl_module_init_order, (jl_value_t*)newm);
+
+    // defer init of children until parent is done being defined
+    // then initialize all in definition-finished order
+    // at build time, don't run them at all (defer for runtime)
+    if (!jl_generating_output()) {
+        if (!ptrhash_has(&jl_current_modules, (void*)newm->parent)) {
+            size_t i, l = jl_array_len(jl_module_init_order);
+            size_t ns = 0;
+            form = (jl_value_t*)jl_alloc_vec_any(0);
+            for (i = 0; i < l; i++) {
+                jl_module_t *m = (jl_module_t*)jl_array_ptr_ref(jl_module_init_order, i);
+                if (jl_is_submodule(m, newm)) {
+                    jl_array_ptr_1d_push((jl_array_t*)form, (jl_value_t*)m);
+                }
+                else if (ns++ != i) {
+                    jl_array_ptr_set(jl_module_init_order, ns - 1, (jl_value_t*)m);
+                }
+            }
+            if (ns < l)
+                jl_array_del_end(jl_module_init_order, l - ns);
+            l = jl_array_len(form);
+            for (i = 0; i < l; i++) {
+                jl_module_t *m = (jl_module_t*)jl_array_ptr_ref(form, i);
                 JL_GC_PROMISE_ROOTED(m);
-                jl_module_load_time_initialize(m);
+                jl_module_run_initializer(m);
             }
-            assert(module_stack.len == l);
-            module_stack.len = stackidx;
-        }
-        JL_CATCH {
-            module_stack.len = stackidx;
-            jl_rethrow();
         }
     }
```

Comment:
For example, the old code would run the `A` method too early in the following highly-contrived example:
```julia
julia> @async @eval module B; sleep(3); __init__() = println("B"); end
Task (runnable) @0x0000000114947190

julia> module A; module A2; __init__() = println("A"); end; sleep(10); end
A
B
```
(should print B after 3 seconds, then A after 10; instead prints A then B after 3 seconds)
This also seems vanishingly-unlikely to make a difference to anyone. I just mention it as a side-effect of the other code changes.